README.md
# OBJECTIVES

The objective behind this repo is to investigate some personal wealth management strategies, from a *theoretical perspective* considering I'm no **F**inancial **P**rofessional **A**dviser ( **FPA** ). **The initial objective** is to empower 
willing individuals with strategies available to *most citizens*, that do not benefit directly from [private wealth management advises](http://en.wikipedia.org/wiki/Wealth_management). 
**The goal** is to *integrate* the knowledge gained from our initial objective into a *custom wealth management plan optimizing one's financial goal(s)*. During that integration process, each strategy 
will be considered as a module with interaction on a main cash flow, so as to allow integration of new strategies in the future, but also to account for differences in strategies applicable to different nations' tax laws (hopefully, 
enough thoughts will be given on that front so that efforts will not be for canadians only).

By means of examples, **and living in Canada**, the *initial investigation* will involved [RRSP](http://en.wikipedia.org/wiki/Registered_Retirement_Savings_Plan) *VS* [TFSA](http://en.wikipedia.org/wiki/Tax-Free_Savings_Account) contributions, 
minimum mortgage payments *VS* anticipated contribution(s) and registered *VS* [nonregistered](http://www.investorwords.com/18431/non_registered_account.html) contributions in light of 
retirement objective(s) and tax-brackets evolution through one's professional life. A potential following *main objective* would be to incorporate all of the previous into an optimization scheme converging toward retirement 
objective(s).

Consequently, given hypothetical investment vehicles returns, the objective is to optimize the flow of cash to such vehicles through various strategies, and for a given financial objective (by default, maximizing retirement income).
 
###### MOTIVATION(s)
I've *rarely* met with FPAs providing an integrated overview of possible **usual** *custom personal financial planning* given a clear picture of current and projected assets along with desired future financial objective(s).
Rather, I've seen spreading (interesting but...) domain specific advises fitting FPAs comfort zones (... and fees criteria). 
*Do ask for a projection of your wealth at age 75, maximizing the retirement income between age 60-75, that optimizes (rather than maximizes) registered/nonregistered contributions along with optimal house mortgage structure, 
and accounting for changes in (1) tax brackets, (2) varying investments performances negative/positive on a yearly basis, (3) family income/tax management ... summarized in a nice graph/table for you to track on a yearly basis !!*

Beyond that statement, as I'm surprised no university research group came up with such framework through some *tax-payed research grant*, I've decided to address the matter and work on it.
(truth is, such research sponsored grants do exist but end up being involved in *technology transfers* that, in the end, will not directly benefit the general public in form of educating its fellow citizen).

... And finally, although I've found interesting information throughout different blogs/books, beside few random or out of context examples, I've never came across actual implementation of strategies ... *and I believe the math !* 
is the real ground on which one may assess the risk-dependency of different financial plans, **in light of FPAs' experiences** in advising different individuals and families while assessing their real ability to 
understand the nature of risk and its impact on one's quality of life.

# INVESTMENT DISCLAIMER, RISKS and WARNINGS
REFER TO [Investment Disclaimer, Risks and Warnings](RisksWarnings.md) FOR DETAILS.

## LICENSE
[ADAPTIVE PUBLIC LICENSE V1.0](http://opensource.org/licenses/alphabetical) as stated in [LICENSE.txt](https://github.com/florentchandelier/WealthManagement/blob/master/License.txt) 
with its supplement file [SUPPFILE.txt](https://github.com/florentchandelier/WealthManagement/blob/master/suppfile.txt).

*Why APL V1.0?* Beside the fact that the initial contributor may make personal choices affecting part of the license terms, **section *3.6* grants independent modules with separate license agreements**. 

Optimistically speaking, this may provide an excellent dynamic for public/private contributions, providing that modularity has been accounted for appropriately (refer to *section 1.7* of APL V1.0) during code design and development.

## CODE LAYOUT and DIRECTORIES

I will use primarily [R](http://www.r-project.org) for implementation as I believe this is a straight-forward yet extremely powerful framework in forms of scripting language for statistics. Along with the code, there should be different **.Rproj**
 files corresponding to **project workspaces** for the excellent [R-Studio IDE](http://www.rstudio.com). **Codes and Scripts validation and Verification was carried on under R-version:>3.0.0 (last test: 3.0.2) and RStudio-version:>0.97.551 (last test: 0.98.484)

Furthermore, **to preserve the confidentiality of personal financial information**, a *fictional financial situation* will be used from sourcing files in the [*FictionalFinancialSituation* directory](https://github.com/florentchandelier/WealthManagement/tree/master/StandaloneTesting/FictionalFinancialSituation), 
containing fictional Income sources, 
house value, mortgage rates and so on. As such, by duplicating such directory into a private one, one may be able to protect his/her financial information while benefiting from, and contributing to the project.

See major changes in the [CHANGELOG](../master/ChangeLog.md) *WIP for automatic generation from git-bash*

### CODE/GIT CONVENTIONS
(eventually, not yet decided: The overall git branching model shall follow the well-illustrated [successful git branching model](http://nvie.com/posts/a-successful-git-branching-model/).)

Branch naming conventions shall follow that of [GroupName/Info](http://stackoverflow.com/questions/273695/git-branch-naming-best-practices):

1. Use **grouping names** at the beginning of your branch names.
2. Define and use short **lead tokens** to differentiate branches in a way that is meaningful to your workflow.
3. Use slashes to separate parts of your branch names.
4. Do not use bare numbers as leading parts.
5. Avoid long descriptive names for long-lived branches.

Grouping Names: Short and well-defined group names (used to tell you to which part of your workflow each branch belongs):

- **wip** Works in progress; stuff I know won't be finished soon
- **feat** Feature I'm adding or expanding
- **bug** Bug fix or experiment
- **junk** Throwaway branch created to experiment
 
### DIR ~ StandaloneTesting
Contains R-scripts for **SINGLE PURPOSE INVESTIGATIONS** of specific investment strategies (such as the Smith Manoeuvre). The objective is to get a feel for different strategies prior further work.
The code itself is meant to be self-explanatory rather than optimized, with web references for definition/explanation/formulae

> REFER TO [Notes_StandaloneTesting](https://github.com/florentchandelier/WealthManagement/blob/master/StandaloneTesting/Notes_StandaloneTesting.md) FOR DETAILS.

### DIR ~ Testing
Contains R-scripts tests for the validation of created function(s) in forms of *(a)* equivalence with expected pre-calculated outcome and *(b)* equivalence with straight forward and anticipated outcome.
> REFER TO COMMENTS WITHIN SOURCES FOR DETAILS

## USAGE
Nothing is implemented yet in the general sense as [Standalone Testing](https://github.com/florentchandelier/WealthManagement/blob/master/StandaloneTesting/Notes_StandaloneTesting.md) of different Strategies is undergoing.



ChangeLog.md
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/2a6409de3f01898786942d177d9b9c42a6c22957) Improving on the ChangeLog script to automatically extract only commits with #changelog in them. The new script is: git log --pretty=format:'[view commit] (http://github.com/florentchandelier/WealthManagement/commit/%H) %s' --first-parent | grep "#changelog" | grep -v Merge > ChangeLog.md #changelog
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/96fb1a500faee4599f36706baba905cd2e855966) aesthetic modification of README.md
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/64bce86f0a5b67f16ccf0f1882c84e1c5112a7fb) Fixes #1 and Fixes #2 new addition(s): - ChangeLog (with automated script) new function(s): - DebtRatioValidation () : rough but a necessary reminder for the future
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/3a74761aeb3fdc6675e43d646afdeddfdddba99b) Create initial ChangeLog.md ; git bash command: git log --pretty=format:'<li> <a href="http://github.com/florentchandelier/WealthManagement/commit/%H">view commit &bull;</a> %s</li>' --first-parent > ChangeLog.md
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/440188fcfb71c37e81aa05163f8a908de00229f3) update Readme with last R and RStudio tested version
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/7e3903c8ba2539237b3689942568bd133495249f) add missing 'library' after 'require' and 'install' to automatically fetch the required libraries at 'source-time'
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/81611a312b7f2e916dd779dcfeaa21c78104055b) *
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/541ec01cf9bddd5b3726c270765027c062e55ad9) reformatting disclaimer
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/276f37fe037ac355eb2eefe0c1e64a5cae1d7289) aesthetic
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/9525153e8e3205c08e40798093772dbc5ceed67c) General: (1) Path modification of the sourced .R scripts for compatibility, (2) use require/install to automatically install required package and (3) update the Smith.md notes
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/bedbbe28a71a805a13f49d05adfa1f209d0dbcf9) Update Notes_StandaloneTesting.md
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/71cd61718576311616987dea249f3e9b8a6c8800) Determining proper Dividends based on tax credits considerations per provinces (Quebec and Ontario).
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/2788da04134c8e741f2aca75e50d62cae78e05ba) Motivations for the Loan-Annuity program
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/409d8d0e2d27bc3e4b5ea808f62bca15bd81d65d) update (wording)
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/cb64b7eef79fd0a0bc083acb0e35c84d415a41bc) Add the CollateralForAnnuity function in the amortize.R module as a first step toward building proper collateral for receiving a "retirement annuity".
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/b4b94b23b4c17859e7271f7261ac3c342be2469d) Merge branch 'master' of github.com:florentchandelier/WealthManagement
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/58d49718f1d1d60a93941979ee595685b4fd1566) correcting http redirection
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/98c2c445d96335a3e0c48bf6cb7a555a56327976) add testing section
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/3d715b467ce26043c120c6d232cebc42711d7c58) NUMBER OF DAYS PER MONTH NOW ACCOUNTED FOR
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/7f922ddcfb293e7038909e6ff7a9fc9aca7ac444) Merge branch 'master' of github.com:florentchandelier/WealthManagement
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/687bbaac7c41986d55879d5b669187dbc4293e48) format correction
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/25dc7f5a6eaaa0bbf6764e17fd2490f2c43ed4a8) damned mark up !
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/39a2ee92ebf9352b2bbbd20a68a5ddf78920751c) * corrections in the readme
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/7374b4a4ee6b0351500384cc3f1acc27534dd8aa) Merge branch 'master' of github.com:florentchandelier/WealthManagement
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/bfca2be2dd39bb74b01a9aa67d0cd58446f73681) updated readme
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/32dc05a8e8655a92cbf1d7f603ce7be4ed7c1ce1) updated readme
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/11a935f39aa7c28ffeaf1f4cdb329d424ae23052) add separate readme for the standalone testing directory
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/2651b20a7b21812e6254d7504ab333dd0130a098) aDDING SEPARATE README NOTE FOR the standalonetesting directory
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/6f3511596483dd214ce35ac16380c9b8d1c8991a) test
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/8bc8f953b71fa46694079a58ff97e7281fc13734) added url for license
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/bb271de935fd57cceac0aa303c29dd8b11616609) Initial description of project and motivation along with Licensing terms
- [view commit](http://github.com/florentchandelier/WealthManagement/commit/3c2be6d93e68ba689dd1e37d348afa3867507052) testing markdown for license

Tests/Test_Smith.R
source("StandaloneTesting/SmithManoeuver.R")
# http://f.redflagdeals.com/showthread.php?t=150279&page=39
#The logic in this statement eludes me. A higher HELOC rate means, ironically, the mortgage is paid off faster, with all 
#else being equal, than a lower rate. But, at the same moment in time, where ever you project it, your networth is greater if 
#you have a lower HELOC rate.
#Try plugging in a $200k mortgage, 12 year amortization, 7% interest rate, 12 payments/year, Step 1 and Step 2, 
#10% compounded investment growth (a la Smith's calculator - that's 10.471307% for the rest of us), Marginal Tax rate of 40%, 
#and then alternate between 6% and 7% HELOC rate.
                                                                                                                                                 
#You will pay the mortgage off 1 month early with the higher HELOC and have $41,183 portfolio net of LOC. With the lower LOC, 
#you will have a mortgage balance of $1,999 but a $51,869 portfolio net of LOC at the same point in the future.
                                                                                                                                                 
# I know what I'd rather have... if you want, I could try to loan you money at 1,000% interest and replace your HELOC. 
#Sounds like it would be a win-win situation. ;)

Test_Smith_1 <- function ()
{
  # THE OJECTIVE IS TO VALIDATE THE FUNCTION WE HAVE DEVELOPPED
  # we verify that, without accounting for the SMG benefits converted back into the Mortgage, 
  # converting the Mortgage at once, is equivalent to converting the Mortgage on a yearly rolling-base
  #
  
  TotalLoan = 476000;# Mortgage Balance
  NAiR=2.5/100; yrs=25; CompPeriod=2; NbYrlyPayment=12; 
  DateLoanOpen = as.Date(format(Sys.Date(), format="%Y-%m-%d"))  
  #
  #  (1)
  #
  Mtg = MortgageStructureInit(DateLoanOpen, TotalLoan);
  
  while ( Mtg$Balance[length(Mtg$Balance)] > 0){
    MtgTemp <- YrlyMortgageSturcture(NAiR, yrs, CompPeriod, NbYrlyPayment, TotalLoan,
                                     Mtg$Balance[length(Mtg$Balance)], 
                                     Mtg$Schedule[length(Mtg$Schedule)]);
    Mtg <- mapply(c, Mtg, MtgTemp, SIMPLIFY=FALSE)  }
  
  SMG_atOnce = SmithManoeuvreGuerillaInit(DateLoanOpen);
  SMG_atOnce<- SmithGuerrilla (Mtg, max(cumsum(SMG_atOnce$P2S)),
                            SMG_atOnce$PeriodicInterest[length(SMG_atOnce$PeriodicInterest)])
  #
  #  (2)
  #  
  MtgSmith = MortgageStructureInit(DateLoanOpen, TotalLoan);
  SMG_yrly = SmithManoeuvreGuerillaInit(DateLoanOpen);
  
  while ( MtgSmith$Balance[length(MtgSmith$Balance)] > 0){
    MtgTemp <- YrlyMortgageSturcture(NAiR, yrs, CompPeriod, NbYrlyPayment, TotalLoan,
                                     MtgSmith$Balance[length(MtgSmith$Balance)], 
                                     MtgSmith$Schedule[length(MtgSmith$Schedule)]);
    
    # We use max(cumsum) as cumsum returns a list for cumsum, the rolling value being the last one.
    SMGTemp <- SmithGuerrilla (MtgTemp, max(cumsum(SMG_yrly$P2S)), 
                               SMG_yrly$PeriodicInterest[length(SMG_yrly$PeriodicInterest)])
    
    MtgSmith <- mapply(c, MtgSmith, MtgTemp, SIMPLIFY=FALSE)
    SMG_yrly <- mapply(c, SMG_yrly, SMGTemp, SIMPLIFY=FALSE)    
  }
  
  if(length(SMG_yrly$P2S) == length(SMG_atOnce$P2S)){cat("Smith Length test Passed \n")} else {cat("Smith Length test FAILED \n")}
  if(max(cumsum(SMG_yrly$P2S-SMG_atOnce$P2S)) < 1){cat("Smith Equivalence test Passed \n")} else {cat("Smith Equivalence test FAILED \n")}
  
}

Test_Smith_2 <- function ()
{
  # THE OJECTIVE IS TO VALIDATE THE FUNCTION WE HAVE DEVELOPPED
  # we verify that, without accounting for the SMG benefits converted back into the Mortgage, 
  # converting the Mortgage at once, is equivalent to converting the Mortgage on a yearly rolling-base
  #
  TotalLoan = 476000;# Mortgage Balance
  NAiR=2.5/100; yrs=25; CompPeriod=2; NbYrlyPayment=12; 
  DateLoanOpen = as.Date(format(Sys.Date(), format="%Y-%m-%d"))  
  #
  #  (1)
  #
  Mtg = MortgageStructureInit(DateLoanOpen, TotalLoan);
  
  while ( Mtg$Balance[length(Mtg$Balance)] > 0){
    MtgTemp <- YrlyMortgageSturcture(NAiR, yrs, CompPeriod, NbYrlyPayment, TotalLoan,
                                     Mtg$Balance[length(Mtg$Balance)], 
                                     Mtg$Schedule[length(Mtg$Schedule)]);
    Mtg <- mapply(c, Mtg, MtgTemp, SIMPLIFY=FALSE)  }
  
  SMG_atOnce = SmithManoeuvreGuerillaInit(DateLoanOpen);
  SMG_atOnce<- SmithGuerrilla (Mtg, max(cumsum(SMG_atOnce$P2S)),
                               SMG_atOnce$PeriodicInterest[length(SMG_atOnce$PeriodicInterest)])
  
  #
  #  (2)
  #  
  
  Total = 0;
  MhtlyNumber = SMG_atOnce$P2S[SMG_atOnce$P2S>0][1] ;
  MthProrated = NAiR*30/365; # assuming 30 days ... only an approximation for quick validation
  NbMonth = yrs*12
  ScheduleCash = rep(MhtlyNumber, NbMonth)
  TotalCashBorrowed = cumsum(ScheduleCash)
  ScheduleInterest = TotalCashBorrowed * MthProrated;
  
  if(abs(ScheduleInterest[length(ScheduleInterest)]*12 - max(TotalCashBorrowed)*NAiR) < 500) 
  {cat("Smith COST FROM HELOC calculation OK")} else {cat("Smith COST FROM HELOC calculation ERROR")}
  
}


Tests/Test_Amortization.R
#
# FROM http://en.wikipedia.org/wiki/Mortgage_calculator
# For example, for a home loan for $200,000 with a fixed yearly interest rate 
# of 6.5% for 30 years : Mthly Mortgage Payment = 1264.14
source("./StandaloneTesting/amortize.R")

Amortize_Test1 <- function ()
{
  
  TotalLoan = 476000;# Mortgage Balance
  NAiR=2.5/100; yrs=25; CompPeriod=1; NbYrlyPayment=12; 
  DateLoanOpen = as.Date(format(Sys.Date(), format="%Y-%m-%d"))
  
  #
  # (1)
  #
  Mtg_atOnce = amortize(TotalLoan, MonthlyAmortization (TotalLoan, NAiR, yrs*NbYrlyPayment),NAiR, yrs*NbYrlyPayment)
  
  
  #
  # (2)
  #
  
  Mtg_yrly = MortgageStructureInit(DateLoanOpen, TotalLoan);
  while ( Mtg_yrly$Balance[length(Mtg_yrly$Balance)] > 0){
    MtgTemp <- YrlyMortgageSturcture(NAiR, yrs, CompPeriod, NbYrlyPayment, TotalLoan,
                                     Mtg_yrly$Balance[length(Mtg_yrly$Balance)], 
                                     Mtg_yrly$Schedule[length(Mtg_yrly$Schedule)]);
    Mtg_yrly <- mapply(c, Mtg_yrly, MtgTemp, SIMPLIFY=FALSE)  }
  
  if (length(Mtg_atOnce$principal) == length(Mtg_yrly$Principal[2:length(Mtg_yrly$Principal)]))
  {paste("Amotization Length test 1 PASSED")} else {paste("Amotization Length test 1 FAILED")}
  
  if (max(cumsum(Mtg_atOnce$principal-cumsum(Mtg_yrly$Principal[2:length(Mtg_yrly$Principal)]))) < 5)
  {paste("Amotization Length test 1 PASSED")} else {paste("Amotization Length test 1 FAILED")}
}

#
# We test the compounding function by validating that the doubling time is equivalent to ln(2)/ln(1+rate)
# as described here: http://en.wikipedia.org/wiki/Rule_of_72
#
Amortize_Test2 <- function ()
{
  amount = 100;
  TimeRatio = 2;
  r = 10/100;
  if (Compounding(amount, r, log(TimeRatio)/log(1+r)) == amount*TimeRatio)
  {paste("Amotization Compounding doubling time test 2 PASSED")} else {paste("Amotization Compounding doubling time test 2 FAILED")}
}

Amortize_Test3 <- function ()
{
  if (round(MonthlyAmortization (200000, 6.5/100, 30*12)) != 1264)
    {paste("Amortize.R->MonthlyAmortization Test 3 FAILED")}else 
    {paste("Amortize.R->MonthlyAmortization Test 3 SUCCEED")}
  
  if (length(which(amortize(200000, MonthlyAmortization (200000, 6.5/100, 30*12),6.5/100, 40*12)$balance>0))+1 != 30*12)
    {paste("Amortize.R -> AMORTIZE SUMMARY Test 3 FAILED")}else 
    {paste("Amortize.R -> AMORTIZE SUMMARY Test 3 SUCCEED")}
}

Amortize_Test4 <- function ()
{
  Borrowed = 150000
  TotalBwd = 0
  yrlRate = 5.5/100
  NbYrs = 25
  
  for (i in 1:NbYrs)
  {    TotalBwd = TotalBwd + Borrowed*((1+yrlRate)^(NbYrs-i+1))  }
  
  if (CollateralForAnnuity() == TotalBwd) {paste("Amortize.R -> Test 4 FAILED on Collateral Assignment")}else 
  {paste("Amortize.R -> Test 4 SUCCEED on Collateral Assignment")}
}

Amortize_Test5 <- function ()
{
  # Interest will be factored in, thus paying interest on interest (providing this is accepted by the lender
  # and mostly based still on the amount of cash value used as collateral from the whole life insurance)
  Annuity <- 150000
  RequiredCollateral <- 0;
  Interest <- 0;
  
  for (i in 1:NbYrs)
  {
    RequiredCollateral <- RequiredCollateral + (Annuity+Interest)*(1+yrlRate);
    Interest <- Interest + (Annuity+Interest)*yrlRate;
  }
  
  ExpectedCollateral <- CollateralForAnnuity(Annuity, yrlRate=5.5/100,NbYrs=25,Loan=FALSE);
  
  if (RequiredCollateral == ExpectedCollateral) {paste("Amortize.R -> Test 5 FAILED on Expected Collateral Annuity")}else 
  {paste("Amortize.R -> Test 4 SUCCEED on Expected Collateral Annuity")}
}

StandaloneTesting/FictionalFinancialSituation/HouseMortgageStructure.R
LoanStructure <- NULL
LoanStructure$HomeValue <- 450000


LoanStructure$CashDown <- 20/100*LoanStructure$HomeValue
LoanStructure$PaymentPercentSalary <- 30/100

LoanStructure$HomeEquityLineOfCredit$AnnuelRate <- 3.5/100 # usually this is variable rates

LoanStructure$Loan$Value <- LoanStructure$HomeValue-LoanStructure$CashDown #300000
LoanStructure$DateLoanOpen <- as.Date(format("2013-12-28", format="%Y-%m-%d"))# NULL # as.Date(format("2010-10-01", format="%Y-%m-%d")) # 2010-10-01
LoanStructure$Loan$AnnualRate <- 3.00/100 # nominal annual interest rate: NAIR
LoanStructure$Loan$Months <- 12 # to be accounted for if using variable rates
LoanStructure$Loan$CompPeriod <- 2; # usually loan compounds semi-annually
LoanStructure$Loan$TermInYrs <- 25; # the number of monthly payments, called the loan's term ;300mth = 25yrs


# LoanStructure$Loan$AnnualRateEvolution <- # Determine expected evolution of Annual Loan Rate

# Should be input: usually MinPayment or MinPayment*2 without additional fees.
LoanStructure$Loan$AnnualReimb <- 0


StandaloneTesting/amortize.R
# TEST: source("./Tests/Test_Amortization.R")

# http://www.acad.polyu.edu.hk/~machanck/lectnotes/c7_finan210.pdf
# Pn = P0(1 + i)^n    ; Pn

if (!require(lubridate)) {install.packages("lubridate")}; library(lubridate)
if (!require(ggplot2)) {install.packages("ggplot2")}; library(ggplot2)
if (!require(scales)) {install.packages("scales")}; library(scales)
if (!require(plyr)) {install.packages("plyr")}; library(plyr)

Compounding <- function (Amount, Rate, NbPeriods, Forward=TRUE)
{
  # output the Amount compounded over the given period at Rate
  if (Forward) return (Amount*(1+Rate)^NbPeriods)
  # output the constant Rate per period required to deliver the compounded_Amount=Rate*Amount over complete period
  else return (exp(log(1+Rate)^1/NbPeriods)-1)
}

InvestmentPerfCompounded <- function (Amount, Rate, NbYearsContrib, NbYearsCompound) {
# res = InvestmentPerfCompounded(15000, 15/100, 11, 65-34); formatC(res, format="d", big.mark=',')
 
  CapTaxGain = 0.79 # http://en.wikipedia.org/wiki/Capital_gains_tax#Canada
  Tot = Amount*(1+Rate)
  for (i in 2:NbYearsCompound)
  {
    if ( i <= NbYearsContrib)
    { Gain = (max(Tot) + Amount)*Rate * CapTaxGain
      Tot = append(Tot, max(Tot)+ Amount+Gain) }
    else
    { Gain = max(Tot)*Rate * CapTaxGain
      Tot = c(Tot, max(Tot) +Gain)}
  }
  return(Tot)
}

MonthlyAmortization  <- function(loan, apr_IN_percent, months) {
    rate <- 1 + apr_IN_percent / 12
    
    # http://en.wikipedia.org/wiki/Amortization_calculator
    # A = P * [i(1+i)^n / (1+i)^n -1]
    # loan * rate^months * (rate - 1) / (rate^months - 1)
    # i = rate-1
    # A = P * [ i+ (i/ ((1+i)^n -1) ) ] as last presented in Wikipedia
    loan * ((rate-1)+((rate-1)/((rate^months)-1) ))
}

MortgageStructureInit <- function (InitialDate=as.Date(format(Sys.Date(), format="%Y-%m-%d")), TotalLoan=20000)
{
  Mtgdata <- list(Schedule=InitialDate, Interest=0, Principal=0, Balance=TotalLoan)
  return(Mtgdata)
}

YrlyMortgageSturcture <- function (NAiR=7.5/100, yrs=5, CompPeriod=2, NbYrlyPayment=12, TotalLoan=20000, balance=20000, PayDay=as.Date(format(Sys.Date(), format="%Y-%m-%d")))
{
  # http://www.vertex42.com/ExcelArticles/amortization-calculation.html
  # Calculating the Rate per Period
  # NAIR = nominal annual interest rate
  # CompPeriod = number of compounding period per year, NbYrlyPayment = number of payment periods per year
  # nbp = number of payment of periods, i = nominal annual interest rate (NAiR)
  
  # if number of compounding period matches number of payment periods then rate per period
  # is nominal rate / 12
  

  RatePerPayment <- ((1+(NAiR/CompPeriod))^(CompPeriod/NbYrlyPayment))-1
  nbp <- NbYrlyPayment*yrs
  amount <- TotalLoan * (RatePerPayment*(1+RatePerPayment)^nbp)/(((1+RatePerPayment)^nbp)-1)
  
  Interest <- NULL; Principal <- NULL;
  Schedule <- NULL; Balance <- NULL
  RemainingBalance <- balance
  
  Init = TRUE
  while (RemainingBalance > 0 && (month(PayDay)<12 || Init) )
  { 
    Init = FALSE
    
    #
    # We assume payments every month; [Get max number of days for new month] and change accordingly
    # - We must pay attention to the number of days per month (otherwise NA will be stored instead of dates); 
    #     . starting on the 29/30 is tricky for the month of february
    #
    month(PayDay) <- month(PayDay) + 1
    
    Schedule <- append(Schedule, PayDay)
    
    Interest <- append(Interest, RemainingBalance * RatePerPayment)
    Principal <- append(Principal, amount - RemainingBalance * RatePerPayment)
    RemainingBalance <- RemainingBalance - (amount - RemainingBalance*RatePerPayment)
    Balance <- append(Balance, RemainingBalance)
  }
  
  Mtgdata <- list(Schedule=Schedule, Interest=Interest, Principal=Principal, Balance=Balance)
  return(Mtgdata)
}

DisplayMortgage <- function(NbComponents, RenderData, Label, XRef, Title, YLegend, XLegend, RenderPoint=NULL){ 
  Plotdata <- data.frame(
    Schedule=rep(XRef,NbComponents), Mortgage=RenderData, 
    Legend=gl(NbComponents,length(XRef),labels=Label)  )

  p <- ggplot(aes(x=Schedule, fill=Legend), data=Plotdata) + geom_line(aes(y= Mortgage, color=Legend))
  
  fmtExpLg10 <- function(x) paste(round_any(x/1000, 0.01) , "K $", sep="")
  p <- p + scale_y_continuous(label=fmtExpLg10)
  p <- p + ggtitle(Title) + ylab(YLegend) + xlab(XLegend)
  print(p)
  
  return(p)
}

amortize <- function(loan, payment, apr_IN_percent, months) {

  rate <- 1 + apr_IN_percent / 12
  month <- 0:months
  
  balance <- loan * rate^month - payment * (rate^month - 1) / (rate - 1)
  complete  <- match(TRUE, balance <= 0) # payment can be more than minimal mortgage payment, thus we look for the first time balance is null
  if(is.na(complete)) complete = length(balance) # if payment is equal to minimal mortage payment, we may be close to 0 but not exactly 0 at term (between 0-1)
  balance <- ifelse(month < month[complete], balance, 0)
  principal <- loan - balance
  interest  <- payment * month - principal
  interest  <- ifelse(month < month[complete], interest, 
                      interest[complete-1])
  

  StartDate = as.Date(format(Sys.Date(), format="%Y-%m-%d"))
  timeline <- seq(StartDate,  length=months ,by="1 month")
  
  amrt <- list(month = month[-1], timeline = timeline, balance=balance[-1], 
               principal = principal[-1], interest = interest[-1],
               paid = principal[-1] + interest[-1], loan=loan,
               payment=payment, apr=apr_IN_percent)
  return(amrt)
  
}

summary.amortization <- function(amrt) {
  cat("  loan amount: ", amrt$loan, "\n")
  cat("          APR: ", amrt$apr_IN_percent, "\n")
  cat("      payment: ", amrt$payment, "\n")
  cat("       months: ", match(TRUE, amrt$balance <= 0), "\n")
  cat("interest paid: ", max(amrt$interest), "\n")
  cat("   total paid: ", max(amrt$paid), "\n")
}

plot.amortization <- function(amrt) {
  year <- amrt$month / 12
  plot(year, amrt$balance, type="l", ylab="Dollars ($)",
       main=paste("Amortization ($", format(amrt$loan, big.mark=","),
                  ", ", sum(amrt$bonus + amrt$anticipated)/amrt$TotalPeriod, "$/yr (boni) + $", format(amrt$AverageMthPay, big.mark=",",
                                                                                                       nsmall=2, digits=2), "/mo)", sep=""), ylim=c(0,max(amrt$paid)),
       xlab="Year")
  lines(year, amrt$interest, col="red", lty=2)
  lines(year, amrt$principal, col="green", lty=2)
  lines(year, amrt$paid, col="blue", lty=2)
  legend(x=1, y=max(amrt$paid),
         legend=c("balance", "interest", "principal", "paid"), 
         col=c("black","red","green","blue"), lty=c(1,2,2,2))
  
  text(floor(max(year)-1),round(max(amrt$paid)), format(round(max(amrt$paid)), scientific = FALSE), col="blue4", font=2)
  text(floor(max(year)-1),round(max(amrt$principal)), format(round(max(amrt$principal)), scientific = FALSE), col="darkgreen", font=2)
  text(floor(max(year)-1),round(max(amrt$interest)), format(round(amrt$interest[length(amrt$interest)]), scientific = FALSE), col="brown4", font=2)
  
}

CollateralForAnnuity <- function (Amount=8094897, yrlRate=5.5/100, NbYrs=25, Loan=TRUE)
{
  # http://finance.zacks.com/collateral-assignment-life-insurance-policy-5962.html
  # OBJECTIVE(S)
  #               - providing a collateral (ColAmount), determine a maximum yrly amount that can be borrowed until
  #                 total consumption of such collateral for a given number of year(s), PAYING INTEREST ON INTEREST;
  #               - Collateral is a LIQUID ASSET, preferred CASH, OR 
  #                       ** considering a whole life policy, its Total Cash Value at age 99 (upon death) ; 
  #                       To use a life insurance policy as collateral for a loan, you must have a 
  #                       policy that accumulates a cash value [http://www.ehow.com/info_7822529_can-life-policy-collateral-loan.html]
  #                       ** Whole life insurance cash value can supplement your resources whilst you are 
  #                          alive by serving as collateral if you require a loan. [http://www.wholelifeinsurance.org.uk/whole-life-insurance-cash-value.html]
  
  
  # r = 1 + yrlRate
  # Collateral = AnnualLoan * (r^(NbYrs-1) + r^(NbYrs-2) ... + r^(NbYrs-(NbYrs-1)))
  #            = Annuity * (1+r) + (Annuity+Interest_yr_1) * (1+r) + (Annuity+Interes_yr_1+2) * (1+r) ........ (compunding the interest on interest)
  #                 , where Interest_yr_1 = Annuity * r ; Interest_yr_1+2 = (Interest_yr_1+Annuity) * r
  #                                                                       = [(Annuity*r)+Annuity] * r
  
  if (Loan)
  {
    # Return the maximal annuity available through the provided the collateral
    # Notice this accounts for INTEREST ON INTEREST
    CollateralAssignment <- Amount
    return( CollateralAssignment/(max(cumsum((1+yrlRate)^(1:NbYrs)))) )
  } else
  {
    # Return the Collateral required for a desired annuity
    # Notice this accounts for INTEREST ON INTEREST on the annuity
    Annuity <- Amount
    return( Annuity * (max(cumsum((1+yrlRate)^(1:NbYrs)))) )
        
  }
}